### PR TITLE
Berlin support

### PR DESCRIPTION
**Pull request by joshuajbouw**
This PR adds both [EIP-2929] and [EIP-2930] into the EVM. 

This is a key PR as it brings the necessary changes to get Berlin hard fork working with the EVM. Since this is a critical update, a discussion and review will be very welcomed. Do note that we will also be updating [evm-tests] as well to ensure full compatibility with Ethereum VM.

- [x] EIP-2929
- [x] EIP-2930

Others related to Berlin, [EIP-2565] and [EIP-2718], should be done on libraries implementing this change.

[EIP-2565]: https://eips.ethereum.org/EIPS/eip-2565 "ModExp Gas Cost"
[EIP-2718]: https://eips.ethereum.org/EIPS/eip-2718 "Typed Transaction Envelope"
[EIP-2929]: https://eips.ethereum.org/EIPS/eip-2929 "Gas cost increases for state access opcodes"
[EIP-2930]: https://eips.ethereum.org/EIPS/eip-2930 "Optional access lists"
[evm-tests]: https://github.com/rust-blockchain/evm-tests

### CODE REVIEW

**Comment by sorpaas** [comment]
Can you merge master and do rustfmt again? Thank you!

**Comment by joshuajbouw** [comment]
This should be good to go now @sorpaas

File: core/src/memory.rs
Comment (sorpaas) [inline core/src/memory.rs:147]: This is actually not necessary for semantics, since below we call `set`, and within that function it checked the len is zero case.

I understand the concern of a short-cut here since we do not need to do additional checks below, but there are indeed several things:

* If we add this check, it should be below the `ErrorFatal::NotSupported` check but not above. The `ErrorFatal` type is a unique error type not covered in spec but useful in practice for non-metered cases -- it ensures we can always handle the addressing spaces since EVM supports 256-bit memory, but in reality all current EVM executors can only support up to 64-bit memory running on x86_64.
* The len zero condition should be triggered extremely rarely since it's basically a no-op. And a well-optimized EVM contract should not have triggered it.
* Our memory code `set` and `copy_large` is already quiet complex and it happened that some trivial changes in the past caused compliance issue. We should really not add additional "patches" unless necessary and instead prefer to fully refactor those two functions.

With all the above said, my current recommendation is to remove this change. We then spend some time refactor `set` and `copy_large` to make it less error-prone and then consider to add those sugar-sized optimizations.

**Comment by sorpaas** [comment]
Please merge master for the jsontests to start!

File: core/src/memory.rs
Comment (birchmd) [reply core/src/memory.rs:147]: I added this change to pass this [eth json test](https://github.com/ethereum/tests/commit/17f7e7a6c64bb878c1b6af9dc8371b46c133e46d) (you can see it linked from [the commit where the change was introduced](https://github.com/rust-blockchain/evm/pull/40/commits/3e081dc934d12d4405c082bcf2ff968ec8cc382d)).

That eth json test was added before Berlin tests so when we updated the evm tests repo to point at a later version of ethereum tests this change was needed to have them all pass.

I agree it's a little hacky, but I do not think we should block merging Berlin changes to refactor the memory functions. We could create an issue to do that refactoring and come back to it after Berlin changes are merged.

Comment (sorpaas) [reply core/src/memory.rs:144]: We don't need that now though -- till the moment we update jsontests.

Alternatively, you can add comments to this line explaining why this is introduced now, and that this is a temporary solution and will eventually be refactored. The priority here should be to avoid more confusions.

Comment (birchmd) [reply core/src/memory.rs:149]: Added a comment in https://github.com/rust-blockchain/evm/pull/40/commits/9bee750378387dd8eaa7bb774547fa89f16d68be

> till the moment we update jsontests.

Are we not going to update the jsontests along side this PR?

File: gasometer/src/lib.rs
Comment (sorpaas) [inline gasometer/src/lib.rs:90]:
```suggestion
	fn spawn(self, gas_limit: u64) -> Self {
```

File: gasometer/src/lib.rs
Comment (sorpaas) [inline gasometer/src/lib.rs:113]:
```suggestion
	pub fn accessed_addresses(&self) -> Option<alloc::collections::btree_set::Iter<'_, H160>> {
```

File: gasometer/src/lib.rs
Comment (sorpaas) [inline gasometer/src/lib.rs:121]:
```suggestion
	pub fn accessed_storages(
```

File: src/executor/stack/mod.rs
Comment (sorpaas) [inline src/executor/stack/mod.rs:139]: What's the rationale for this change? One of the drawback of this is that it'll be potentially difficult for users to get a single type for `StackExecutor` suitable for different configurations of precompiles. Same reason why `Config` is not a generic parameter.

File: gasometer/src/lib.rs
Comment (sorpaas) [inline gasometer/src/lib.rs:548]: The list of `accessed_addresses` and `accessed_storages` really shouldn't belong to `gasometer`. A better design is to create a separate struct `Accessed` specifically to hold them, and add an extra field in executor metadata. This line can then be changed to a `is_cold: bool` which we continue to handle the access gas.

The EVM step cycle has two phrases, one `immutable` where we gather necessary data such as gas cost, to determine if we should continue the execute, and another `mutable` where we actually do the execution. By putting access list to gasometer this invariant is broken (which is made apparent by the below change of an immutable reference to a mutable one).

* `immutable` peek the access list, determine if the address is cold.
* `immutable` pass the coldness along with other information to gasometer and get the gas costs.
* `mutable` do the execution / update the gas metadata and access list.

The above invariant may not sound that useful if it's only about a full execution cycle, but it is extremely relevant if we want to use this crate to build a debugger or a stepper.

Comment (birchmd) [reply gasometer/src/lib.rs:548]: I had noticed this pattern and my first attempt at this functionality went in this direction. You can see it in [this commit](https://github.com/rust-blockchain/evm/pull/40/commits/d7ea7678b288f9461949f1ac686653f649812829#diff-bf2551b9edd7333cb28398048e01898cfd8c41877603ff3e17ea444fe8cf6e51R784).

What I didn't like about this approach is the need to thread the `is_cold` functionality through the `Handler` because that was the only way for the gasomete to get access to the wider state. But that felt off to me because it is only the gasometer that cares about these access lists, so I thought it might be better to encapsulate inside the gasometer itself.

But I respect if you think maintaining this strict mutability flow is more important. Just let me know and I can change it back to my first version.

Comment (sorpaas) [reply gasometer/src/lib.rs:548]: Yeah I do think that version would be better. `Handler::is_cold` is okay -- we do that for `Handler::exists` as well (which is kind of also EIP-specific).

Comment (birchmd) [reply gasometer/src/lib.rs:548]: Sounds good, thanks for the clarification. I'm OOO tomorrow, but I'll push a fix for this on Monday.

File: src/executor/stack/mod.rs
Comment (joshuajbouw) [reply src/executor/stack/mod.rs:139]: That might've been an overzealous mistake there. I'll double check.

File: core/src/memory.rs
Comment (joshuajbouw) [reply core/src/memory.rs:149]: That's what I am confused about, as jsontests will be updated immediately after.

File: gasometer/src/lib.rs
Comment (joshuajbouw) [reply gasometer/src/lib.rs:548]: I can get this is before Monday.

Comment (joshuajbouw) [reply gasometer/src/lib.rs]: Here: https://github.com/rust-blockchain/evm/pull/40/commits/c2af4516d210b7520f6cb1c4c5d2d16b6a9d6103

File: core/src/utils.rs
Comment (sorpaas) [inline core/src/utils.rs:9]: Is this also related to clippy errors?

Comment (birchmd) [reply core/src/utils.rs:9]: Probably https://rust-lang.github.io/rust-clippy/master/index.html#enum_variant_names

Comment (joshuajbouw) [reply core/src/utils.rs:9]: What @birchmd. Just something that clippy was catching.

Comment (birchmd) [reply core/src/utils.rs:9]: @joshuajbouw I know. @sorpaas asked about it above and I was taking a guess at what the specific clippy warning might have been.

File: src/executor/stack/mod.rs
Comment (joshuajbouw) [reply src/executor/stack/mod.rs:211]: @sorpaas I guess that I misunderstood. So, if I change it so that every time we use `call` and `create`, I could pass in the precompile set there but, that may be a bit cumbersome.

Another option would be to make a struct called `Precompiles`, over a trait, which you then push every single `fn` and other data...?

The issue is for one of the EIPs for Berlin, we need to know the addresses for the precompiles so this change is absolutely necessary.

Comment (sorpaas) [reply src/executor/stack/mod.rs:211]: That should be fine. I haven't found which EIP you're referring to but we can just add the parameter to the precompile function if needed.

What I really don't want to happen, however, is that the precompile type being dependent on executor type, like having the executor state type passed to precompile. In that case we get circle type dependency.

Comment (sorpaas) [reply src/executor/stack/mod.rs:211]: I think you can just change the precompile interface to the following to avoid the generic.

```
pub type Precompile =
	(H160, fn(&[u8], Option<u64>, &Context) -> Option<Result<PrecompileOutput, ExitError>>);
```

Or better -- make it a struct to be more descriptive.

Comment (joshuajbouw) [reply src/executor/stack/mod.rs]: This should do it: https://github.com/rust-blockchain/evm/pull/40/commits/5bad029315c35c77841b7a457c2a59ddcf370db2

**Comment by joshuajbouw** [comment]
Hold on, getting issues with `into_keys` when integrating this due to it being bleeding edge API.

File: src/executor/stack/mod.rs
Comment (sorpaas) [inline src/executor/stack/mod.rs:116]: Can we add this function back?

**Review by sorpaas — APPROVED** [review]
LGTM mostly, but it would be great to add `StackExecutor::new` back.

**Pull request merged**
